### PR TITLE
Add MQTT module for esp32

### LINKED ISSUE #1946: mqtt module for esp32?
Just wonder when can we have mqtt feature for esp32? very looking forwards to this.

### PR DESCRIPTION
**Pull request by goughy**
Fixes #1946.

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/en/*`.

This pull request adds support for the mqtt protocol and is based on https://github.com/tuanpmt/espmqtt as suggested in issue #1946.  This implementation simplifies mqtt support over the AWS IOT library implementation and has tried to maintain compatibility with the existing NodeMCU MQTT module.  Where the underlying library does not provide support (eg. table-based unsubscribe function), it has been noted in the doc.

Secure (TLS) MQTT support has not been tested (in progress), but there is no reason it should not work out of the box.

### CODE REVIEW

File: components/modules/mqtt.c [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+	}
+
+	if( lua_isnumber(L, 5) )
+	{
+		mqtt_cfg->disable_clean_session = (luaL_checknumber( L, 5 ) == 0);
+		n++;
+	}
+	lua_pop( L, n ); //remove parameters
+
+	lua_newtable( L );
```
Comment (devsaurus) [inline components/modules/mqtt.c]: I suggest to build the client's metatable not programmatically but from a rotable, e.g. like in the [`ws2812 module`](https://github.com/nodemcu/nodemcu-firmware/blob/dev-esp32/components/modules/ws2812.c#L500).

File: docs/en/modules/mqtt.md [hunk 2]
```diff
@@ -0,0 +1,266 @@
+# MQTT Module
```
Comment (devsaurus) [inline docs/en/modules/mqtt.md]: `mkdocs.yml` is missing a reference to this file.

File: components/modules/mqtt.c (hunk 1 above)
Comment (goughy) [reply components/modules/mqtt.c]: @devsaurus sure thing.  I will look into it.

**Comment by goughy** [comment]
Thanks for the review.  I will try and find some time in the next few days to modify and test and update the PR.

Cheers,

**Pull request merged**